cpp/src/reductions/reductions.cu

```diff
@@ -203,22 +205,44 @@ struct DeviceMax {
 
 template <typename Op>
 struct ReduceDispatcher {
-    template <typename T,
-              typename std::enable_if_t<std::is_arithmetic<T>::value>* = nullptr>
-    gdf_error operator()(gdf_column *col, 
+    template <typename T>
+    gdf_error launch(gdf_column *col,
                          void *dev_result, 
-                         gdf_size_type dev_result_size) {
+                         gdf_size_type dev_result_size)
+    {
         GDF_REQUIRE(col->size > col->null_count, GDF_DATASET_EMPTY);
         T identity = Op::template identity<T>();
         return ReduceOp<T, Op>::launch(col, identity, 
-                                       reinterpret_cast<T*>(dev_result), 
+                                       static_cast<T*>(dev_result),
                                        dev_result_size); 
     }
 
     template <typename T,
-              typename std::enable_if_t<!std::is_arithmetic<T>::value, T>* = nullptr>
-    gdf_error operator()(gdf_column *col, 
-                         void *dev_result, 
+              typename std::enable_if<std::is_arithmetic<T>::value>::type* = nullptr>
+    gdf_error operator()(gdf_column *col,
+                         void *dev_result,
+                         gdf_size_type dev_result_size)
+    {
+        return launch<T>(col, dev_result, dev_result_size);
+    }
+
+    template <typename T, typename std::enable_if<
+              !std::is_arithmetic<T>::value &&
+              std::is_base_of<DeviceForNonArithmetic, Op>::value
```

Comment:
I think it'd be more consistent with modern C++ design to add a trait or a `constexpr` function called `is_nonarithmetic_op` that returns true for `Max` and `Min`, but false for all the other operators.

Reply:
Yes, both are feasible.
I prefer later case, the implementation will be: 
```
template <typename Op>
struct ReduceDispatcher {
   static constexpr bool is_nonarithmetic_op = std::is_same<Op, DeviceMin>::value || std::is_same<Op, DeviceMax>::value ;

   template <typename T, typename std::enable_if<!std::is_arithmetic<T>::value && is_nonarithmetic_op > >::type* = nullptr> ...
}
```

Reply:
It has been fixed:

https://github.com/rapidsai/cudf/blob/e012acb6085f99d805b3e0399d2419b2239707cc/cpp/src/reductions/reductions.cu#L93-L96